executor/sort.go

```diff
@@ -119,85 +126,133 @@ func (e *SortExec) Open(ctx context.Context) error {
 	}
 	e.exceeded = 0
 	e.spilled = 0
-	e.rowChunksInDisk = nil
-	e.rowPtrsInDisk = e.rowPtrsInDisk[:0]
 	e.partitionList = e.partitionList[:0]
-	e.partitionRowPtrs = e.partitionRowPtrs[:0]
 	return e.children[0].Open(ctx)
 }
 
 // Next implements the Executor Next interface.
+// Sort constructs the result following these step:
+// 1. Read as mush as rows into memory.
+// 2. If memory quota is triggered, sort these rows in memory and put them into disk as partition 1, then reset
+//    the memory quota trigger and return to step 1
+// 3. If memory quota is not triggered and child is consumed, sort these rows in memory as partition N.
+// 4. Merge sort if the count of partitions is larger than 1. If there is only one partition in step 4, it works
+//    just like in-memory sort before.
 func (e *SortExec) Next(ctx context.Context, req *chunk.Chunk) error {
 	req.Reset()
 	if !e.fetched {
+		e.initCompareFuncs()
+		e.buildKeyColumns()
 		err := e.fetchRowChunks(ctx)
 		if err != nil {
 			return err
 		}
-		if e.alreadySpilled() {
-			err = e.prepareExternalSorting()
-			if err != nil {
-				return err
-			}
-			e.fetched = true
-		} else {
-			e.initPointers()
-			e.initCompareFuncs()
-			e.buildKeyColumns()
-			sort.Slice(e.rowPtrs, e.keyColumnsLess)
-			e.fetched = true
-		}
+		e.initPointers()
+		sort.Slice(e.rowPtrs, e.keyColumnsLess)
+		e.fetched = true
 	}
 
-	if e.alreadySpilled() {
-		for !req.IsFull() && e.Idx < len(e.partitionRowPtrs[0]) {
-			rowPtr := e.partitionRowPtrs[0][e.Idx]
-			row, err := e.partitionList[0].GetRow(rowPtr)
-			if err != nil {
-				return err
-			}
-			req.AppendRow(row)
-			e.Idx++
+	if len(e.partitionList) > 0 {
+		if err := e.externalSorting(req); err != nil {
+			return err
 		}
 	} else {
 		for !req.IsFull() && e.Idx < len(e.rowPtrs) {
 			rowPtr := e.rowPtrs[e.Idx]
-			req.AppendRow(e.rowChunks.GetRow(rowPtr))
+			row, err := e.rowChunks.GetRow(rowPtr)
+			if err != nil {
+				return err
+			}
+			req.AppendRow(row)
 			e.Idx++
 		}
 	}
 	return nil
 }
 
-func (e *SortExec) prepareExternalSorting() (err error) {
-	e.initCompareFuncs()
-	e.buildKeyColumns()
-	e.rowPtrsInDisk = e.initPointersForListInDisk(e.rowChunksInDisk)
-	// partition sort
-	// Now only have one partition.
-	// The partition will be adjusted in the next pr.
-	err = e.readPartition(e.rowChunksInDisk, e.rowPtrsInDisk)
-	if err != nil {
-		return err
-	}
+func (e *SortExec) generatePartition() {
 	e.initPointers()
 	sort.Slice(e.rowPtrs, e.keyColumnsLess)
-	listInDisk, err := e.spillToDiskByRowPtr()
-	if err != nil {
-		return err
+	e.partitionList = append(e.partitionList, e.rowChunks)
+	e.partitionRowPtrs = append(e.partitionRowPtrs, e.rowPtrs)
+}
+
+type topNChunkHeapWithIndex struct {
+	*SortExec
```

Comment:
It would be better not embed the entire `SortExec`, but only what it needs.

Reply:
Ok, I will address comment.